File: 460. LFU Cache/Solution.java
class LFUCache {
    Map<Integer, Integer> tracker;
    Map<Integer, Queue<Integer>> freqQueues;
    int[] counter;
    int maxSize, currSize, lowFreq;

    public LFUCache(int capacity) {
        tracker = new HashMap<>();
        freqQueues = new HashMap<>();
        freqQueues.put(1, new LinkedList<>());
        counter = new int[100001];
        maxSize = capacity;
        currSize = 0;
        lowFreq = 1;
    }
    
    public int get(int key) {
        if (!tracker.containsKey(key)) return -1;
        counter[key]++;
        if (!freqQueues.containsKey(counter[key])) freqQueues.put(counter[key], new LinkedList<>());
        freqQueues.get(counter[key]).offer(key);
        return tracker.get(key);
    }
    
    public void put(int key, int value) {
        if (!tracker.containsKey(key)) {
            if (maxSize == 0) return;
            if (currSize == maxSize) {
                boolean found = false;
                while (!found) {
                    Queue<Integer> curr = freqQueues.get(lowFreq);
                    while (!found && !curr.isEmpty()) {
                        int candidate = curr.poll();
                        if (counter[candidate] != lowFreq) continue;
                        tracker.remove(candidate);
                        counter[candidate] = 0;
                        found = true;
                    }
                    if (!found) lowFreq++;
                }
            } else {
                currSize++;
            }
            lowFreq = 1;
        }
        counter[key]++;
        if (!freqQueues.containsKey(counter[key])) freqQueues.put(counter[key], new LinkedList<>());
        freqQueues.get(counter[key]).offer(key);
        tracker.put(key, value);
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 460. LFU Cache/Solution.py
class LFUCache:

    def __init__(self, capacity: int):
        self.tracker = dict()
        self.freqQueues = dict()
        self.counter = [0 for i in range(100001)]
        self.freqQueues[1] = []
        self.lowFreq = 1
        self.currSize = 0
        self.maxSize = capacity

    def get(self, key: int) -> int:
        if key not in self.tracker: return -1
        self.counter[key] += 1
        if self.counter[key] not in self.freqQueues: self.freqQueues[self.counter[key]] = []
        self.freqQueues[self.counter[key]].append(key)
        return self.tracker[key]

    def put(self, key: int, value: int) -> None:
        if key not in self.tracker:
            if self.maxSize == 0: return
            if self.currSize == self.maxSize:
                found = False
                while not found:
                    curr = self.freqQueues[self.lowFreq]
                    while not found and curr:
                        candidate = curr.pop(0)
                        if self.counter[candidate] != self.lowFreq: continue
                        del self.tracker[candidate]
                        self.counter[candidate] = 0
                        found = True
                    if not found: self.lowFreq += 1
            else:
                self.currSize += 1
            self.lowFreq = 1
        self.counter[key] += 1
        if self.counter[key] not in self.freqQueues: self.freqQueues[self.counter[key]] = []
        self.freqQueues[self.counter[key]].append(key)
        self.tracker[key] = value

# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)